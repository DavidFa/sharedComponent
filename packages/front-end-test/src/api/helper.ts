export const HOST_URL = "https://jsonplaceholder.typicode.com12222";

interface ApiError {
    code: number;
    error: string;
}

const isApiError = (x: any): x is ApiError => {
    return typeof x.code === 'number';
};

// generator function sync supar
export const PostApi = async <T, R>(url: string, post: T): Promise<R> => {
    try {
        const response = await fetch(`${HOST_URL}${url}`, { method: "POST", body: JSON.stringify(post) });
        if (!response.ok) {
            // Return the known error for future handling
            // return rejectWithValue({ errorMessage: "error" } as MyKnownError)
            throw new Error("error");
        }
        const data = (await response.json());
        return data;
    } catch (e) {
        throw e;
    }
}

