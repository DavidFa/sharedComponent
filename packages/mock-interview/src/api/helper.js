
export const cats = [
    { id: 1, color: 'white', age: 1 },
    { id: 2, color: 'block', age: 2 },
    { id: 3, color: 'white', age: 3 },
]
export const mockFetchCats = (isSuccess = true) => {
    return new Promise(function (resolve, reject) {
        setTimeout(() => {
            if (isSuccess) {
                return resolve({
                    ok: true,
                    code: 200,
                    json: async () => cats
                });
            } else {
                reject("Netcork Error");
            }
        }, 1000);

    })
}