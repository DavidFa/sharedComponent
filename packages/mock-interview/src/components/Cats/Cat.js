import React, { useState, useCallback } from 'react';
import './Cats.css'


const Cat = ({ id, color, age, onEditSubmitHandler }) => {

    const [cat, setCat] = useState({ id, color, age })

    const catEditHandler = useCallback(
        (e) => {
            const key = e.target.name;
            const val = e.target.value;
            const newCat = { ...cat };
            newCat[key] = val;
            setCat(newCat);
        },
        [cat],
    )

    return (
        <div className="Cat">
            <form className="Form" onSubmit={(e) => onEditSubmitHandler(cat, e)}>
                <div>
                    <label htmlFor="catId">ID:</label>
                    <input id="catId" name="id" value={cat.id} disabled={true} />
                </div>
                <div>
                    <label htmlFor="catColor">Color:</label>
                    <input id="catColor" name="color" value={cat.color} onChange={catEditHandler} />
                </div>
                <div>
                    <label htmlFor="catAge">Age:</label>
                    <input id="catAge" name="age" value={cat.age} onChange={catEditHandler} />
                </div>
                <div className="Button">
                    <button>Submit</button>
                </div>
            </form>
        </div>
    )
}

export default Cat;
