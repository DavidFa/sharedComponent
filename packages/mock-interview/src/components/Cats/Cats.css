.Cats {
    box-sizing: border-box;
    width: 500px;
    margin: 1rem auto;
    background-color: #eee;
    border-radius: 4px;
    border: 1px solid #ccc;
    box-shadow: 4px 4px 40px #ddd;
    padding: 1rem;
}

.Cat {
    box-sizing: border-box;
    width: 500px;
    margin: 1rem auto;
    background-color: #eee;
    border-radius: 4px;
    border: 1px solid #ccc;
    box-shadow: 4px 4px 40px #ddd;
    padding: 1rem;
}

.Form {
    width: 434px;
    display: flex;
    flex-direction: column;
}

.Form>div {
    width: inherit;
    margin: 3px;
}

.Form>div>label {
    width: 100px;
    text-align: right;
    padding: 5px 10px;
    display: inline-block;
}

.Form>div>input {
    width: 300px;
    text-align: left;
    padding: 5px 5px;
    display: inline-block;
}

div.Button {
    text-align: right;
}

div.Button button{
    background-color: rgb(87, 145, 233);
    border-radius: 3px;
    padding: 4px 8px;
    color: white;
    font-weight: bold;
}

span {
    margin: 3px 6px;
}