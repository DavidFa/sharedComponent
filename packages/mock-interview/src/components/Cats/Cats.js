import { useCats } from './hooks/useCats';
import './Cats.css'


export default function Cats() {

    const { cats, editCatHandler, getCat } = useCats();

    return (
        <>
            {getCat()}
            <div className="Cats" data-testid="wrapper">
                {cats.catList.length > 1 && cats.catList.map(cat => <div key={cat.id} onClick={() => editCatHandler(cat.id)}>
                    <span>{cat.id}</span>
                    <span style={{ color: 'red' }}>{cat.color}</span>
                    <span>{cat.age}</span>
                </div>)}
            </div>
        </>
    )
}
