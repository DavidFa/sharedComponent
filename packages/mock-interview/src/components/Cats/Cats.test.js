import Cats from './Cats';
import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import { cats } from '../../api/helper';

describe("test useCats hook", () => {
    const mockSetState = jest.fn();

    beforeEach(() => {
        jest.spyOn(React, 'useState').mockImplementation(() => [{ catList: cats, cat: null }, mockSetState])
        // render(<Cats />);
    })

    it('should render cats with given state', () => {
        const {container, getByText} = render(<Cats />)
        const wrapper = screen.queryByTestId(/wrapper/i);
        expect(wrapper).toBeInTheDocument();
        expect(container).toMatchSnapshot();

        // fireEvent.click()
    })

})