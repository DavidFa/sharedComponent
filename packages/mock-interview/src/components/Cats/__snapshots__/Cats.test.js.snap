// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test useCats hook should render cats with given state 1`] = `
<div>
  <div
    class="Cats"
    data-testid="wrapper"
  >
    <div>
      <span>
        1
      </span>
      <span
        style="color: red;"
      >
        white
      </span>
      <span>
        1
      </span>
    </div>
    <div>
      <span>
        2
      </span>
      <span
        style="color: red;"
      >
        block
      </span>
      <span>
        2
      </span>
    </div>
    <div>
      <span>
        3
      </span>
      <span
        style="color: red;"
      >
        white
      </span>
      <span>
        3
      </span>
    </div>
  </div>
</div>
`;
